c2rust-analyze/src/known_fn.rs

```diff
@@ -602,28 +622,823 @@ pub const fn all_known_fns() -> &'static [KnownFn] {
+            ) -> c_int;
+
+            #[cfg(target_os = "linux")]
+            fn inotify_add_watch(
+                fd: c_int,
+                path: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                mask: u32,
+            ) -> c_int;
+
+            #[cfg(target_os = "linux")]
+            fn inotify_init() -> c_int;
+
+            #[cfg(target_os = "linux")]
+            fn inotify_init1(
+                flags: c_int,
+            ) -> c_int;
+
+            #[cfg(target_os = "linux")]
+            fn inotify_rm_watch(
+                fd: c_int,
+                wd: c_int,
+            ) -> c_int;
+
+            // fn ioctl(
+            //     fd: c_int,
+            //     request: c_ulong,
+                   // TODO(kkysen) varargs
+            //     ...
+            // ) -> c_int;
+
+            fn kill(
+                pid: pid_t,
+                sig: c_int,
+            ) -> c_int;
+
+            fn listen(
+                socket: c_int,
+                backlog: c_int,
+            ) -> c_int;
+
+            fn localtime_r(
+                time_p: *const time_t: [READ | NON_NULL],
+                // `READ` because it's returned (if there's no error).
+                result: *mut tm: [READ | WRITE | NON_NULL],
+            ) -> *mut tm: [READ | WRITE];
+
+            fn lseek(
+                fd: c_int,
+                offset: off_t,
+                whence: c_int,
+            ) -> off_t;
+
+            fn lstat(
+                path: *const c_char: [READ | OFFSET_ADD | NON_NULL],
+                buf: *mut stat: [WRITE | NON_NULL],
+            ) -> c_int;
+
+            fn malloc(
+                size: size_t,
+            ) -> *mut c_void: [READ | WRITE | OFFSET_ADD | FREE]; // TODO(kkysen) OFFSET_ADD?
+
+            fn memchr(
+                // `WRITE` because the return type is derived from `cx`'s provenance.
+                cx: *const c_void: [READ | WRITE | OFFSET_ADD | NON_NULL],
+                c: c_int,
+                n: size_t,
+            ) -> *mut c_void: [READ | WRITE | OFFSET_ADD | OFFSET_SUB];
+
+            fn memcmp(
+                cx: *const c_void: [READ | OFFSET_ADD | NON_NULL],
+                ct: *const c_void: [READ | OFFSET_ADD | NON_NULL],
+                n: size_t,
+            ) -> c_int;
+
+            fn memcpy(
+                dest: *mut c_void: [WRITE | OFFSET_ADD | NON_NULL],
+                src: *const c_void: [READ | OFFSET_ADD | NON_NULL],
+                n: size_t,
+                // Same as `dest`, but while `dest` shouldn't be `NULL`,
+                // it can without error, and thus this can return `NULL`.
+            ) -> *mut c_void: [WRITE | OFFSET_ADD];
+
+            fn memmove(
+                // TODO(kkysen) Note: A `&mut []` for `dest` and `&[]` for `src` would be UB since they can overlap.
+                dest: *mut c_void: [WRITE | OFFSET_ADD | NON_NULL],
+                src: *const c_void: [READ | OFFSET_ADD | NON_NULL],
```

Comment:
I don't think this is actually a problem.  We can't safely produce overlapping `&mut [T]` and `&[T]`, so we would instead end up with two `&[Cell<T>]` references in possibly-overlapping cases.

Reply:
Ah, that makes more sense.  I'll add this as a comment, too.